cache/estargz.go

```diff
@@ -11,76 +11,117 @@ import (
 	"github.com/containerd/stargz-snapshotter/estargz"
 	"github.com/moby/buildkit/util/compression"
 	digest "github.com/opencontainers/go-digest"
+	ocispecs "github.com/opencontainers/image-spec/specs-go/v1"
 	"github.com/pkg/errors"
 )
 
 var eStargzAnnotations = []string{estargz.TOCJSONDigestAnnotation, estargz.StoreUncompressedSizeAnnotation}
 
-// writeEStargz writes the passed blobs stream as an eStargz-compressed blob.
-// getAnnotations function returns all eStargz annotations.
-func writeEStargz() (compressorFunc compressor, finalize func(context.Context, content.Store) (map[string]string, error)) {
-	annotations := make(map[string]string)
-	var bInfo blobInfo
+// compressEStargz writes the passed blobs stream as an eStargz-compressed blob.
+// finalize function finalizes the written blob metadata and returns all eStargz annotations.
+func compressEStargz() (compressorFunc compressor, finalize func(context.Context, content.Store) (map[string]string, error)) {
+	var cInfo *compressionInfo
+	var writeErr error
 	var mu sync.Mutex
 	return func(dest io.Writer, requiredMediaType string) (io.WriteCloser, error) {
 			if compression.FromMediaType(requiredMediaType) != compression.Gzip {
-				return nil, errors.Errorf("unsupported media type for estargz compressor %q", requiredMediaType)
+				return nil, fmt.Errorf("unsupported media type for estargz compressor %q", requiredMediaType)
 			}
 			done := make(chan struct{})
 			pr, pw := io.Pipe()
-			go func() {
+			go func() (retErr error) {
 				defer close(done)
-				defer pr.Close()
-				cw, bInfoCh := calculateBlob()
-				defer cw.Close()
-				w := estargz.NewWriter(io.MultiWriter(dest, cw))
-				if err := w.AppendTar(pr); err != nil {
+				defer func() {
+					if retErr != nil {
+						mu.Lock()
+						writeErr = retErr
+						mu.Unlock()
+					}
+				}()
+
+				blobInfoW, bInfoCh := calculateBlobInfo()
+				defer blobInfoW.Close()
+				w := estargz.NewWriter(io.MultiWriter(dest, blobInfoW))
+
+				// Using lossless API here to make sure that decompressEStargz provides the exact
+				// same tar as the original.
+				//
+				// Note that we don't support eStragz compression for tar that contains a file named
+				// `stargz.index.json` because we cannot create eStargz in loseless way for such blob
+				// (we must overwrite stargz.index.json file).
+				if err := w.AppendTarLossLess(pr); err != nil {
```

Comment:
Bit worried that other tools can still create unsupported tars. Can that API be removed? What is the behavior  when such tarballs are pulled? Maybe they should fail the `isEstargz` check?

Reply:
The cases where the unsupported tar (= contains `stargz.index.json`) reach here will be the following

- invalid eStargz: the source blob is a broken (or malformed) eStargz that contains `stargz.index.json` but has an invalid footer. In this case, the blob is recognized as a normal gzip blob and `decompressEStargz` is not applied.
- name conflict: the source blob is a non-eStargz blob (tar/gzip/zstd) but contains `stargz.index.json` which is not TOC.

(If the source blob is an eStargz, the decompressor (`decompressEStargz`) removes the TOC entry before appending so `AppendTarLossLess` won't return the error.)

I think these cases are rare thus it's reasonable to return an error here.

Reply:
Yes, I didn't think the error case but just that `AppendTarLossLess` is an additional API and `AppendTar` still remains. So other tools can create estargz that does not decompress properly.  Can we make this requirement for stargz lib, and what is the behavior when we pull such blob that was not created with a lossless encoder.

Reply:
@tonistiigi 

> other tools can create estargz that does not decompress properly.

If the client converts a plain-gzip image A to eStargz B on their side with tools other than BuildKit, B doesn't hold the original bits so it's treated as a different image than A when pulled to BuildKit. When BuildKit converts B into a tar image, the produced tar is different from what is contained in A.

> what is the behavior when we pull such blob that was not created with a lossless encoder.

Could you elaborate on the problem of the conversion happening outside of BuildKit?  When it comes to tools other than BuildKit, they mostly need to change tar bits for enabling optimization which sorts tar entries in priority order and adds some landmark dummy files. I think BuildKit cache should treat such blobs as different one from the original.

Reply:
Is there any reason other tools couldn't just also use the lossless method. Why can't this be default?

Is it still correct that `isEstargz` detects lossy estargz as estargz or should it use it as regular gzip and try to recompress with lossless method?